### ISSUE TITLE
Constant function literals

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1048]
Static functions are compile time constants, but function literals are not.
It would be great if it was possible to write compile time constant function literals. They would necessarily have static scope (no access to &quot;this&quot;) even if written inside a class declaration.

Constant function literals would create only one function object per function expression (no unification of different function expressions, even if they happen to have the same body). 

This is particularly useful for default function parameters.

I suggest `const <functionLiteral>` as syntax. That would allow you to write, e.g.,:
```dart
class Map<K,V> {
  V getIfPresent(K key, [V onNotPresent() = const () => null]);
}
```
The current alternative is to declare the function as static:
```dart
class Map<K,V> {
  V getIfPresent(K key, [V onNotPresent() = _returnNull]);
  static _returnNull() => null;
}
```

You can use it to ensure that you don't create too many closures if you want to create a function inside a loop:
```dart
for (var x in xs) {
  x.somethingOrError(const (x) => throw new Failure(x));
}
```
which would otherwise create a new closure per iteration of the loop. The alternative is again to define a static function somewhere and use that, but it moves the logic away from where it's most readable.

The syntax shouldn't collide with any current use of &quot;const&quot;, since we currently don't allow that syntax in expression position. It does collide with a constant constructor declaration, but the uses are different enough that I don't think it's a problem. If we want parity, we could allow `const foo() => xx` as a top-level declaration too, meaning the same as `static`, but that would make `static` redundant.

**Comment by tomasdev** [comment]
What is the solution these days for having optional methods?

**Comment by munificent** [comment]
> ```dart
> onPressed: onPressed ?? () {},
```

Yup, that works fine.

**Comment by lukepighetti** [comment]
Not that it matters, but i have a preference for `onPressed: ()=>onPressed?.call()`

**Comment by mateusfccp** [comment]
> Not that it matters, but i have a preference for `onPressed: ()=>onPressed?.call()`

If that's the case, why not `onPressed: onPressed?.call`?

**Comment by lukepighetti** [comment]
That would work too! But the pattern above is the one that I have a preference for. Often times the arguments don't match and I prefer this

```dart
onTap: ()=>onTap?.call(),
onTapUp: (_)=>onTapUp?.call(),
```

over


```dart
onTap: onTap?.call,
onTapUp: (_)=>onTapUp?.call(),
```

**Comment by lrhn** [comment]
Using `ontap?.call` is equivalent to just using `ontap`, so there is no advantage in doing it that way.

The `onPressed?.call` expression can be `null`, whereas `() => onPressed?.call()` is always a function which may or may not do anything. In this context, we're looking for a default value to avoid being `null`, so that's also important.

**Comment by jodinathan** [comment]
with this it would be able to use anonymous functions on annotations, right?

```dart

class Foo {
  final Function fn;
  
  const Foo(this.fn);
}

@Foo(const () {})
class Bar {
  
}
```

**Comment by munificent** [comment]
Yes, if we made some function literals be valid const expressions, then you could use them as arguments to metadata annotations too.

**Comment by ernesto** [comment]
Waiting for the resolution in the "Constant function literals Shippuden"

**Comment by 5hanth** [comment]
+1

**Comment by munificent** [comment]
Hi, @5hanth, I've deleted your "+1" comment because we prefer users use the emoji buttons at the top of the issue to register feedback like that. Doing so makes it easier for use to compare feedback across issues and declutters the comment thread. We do appreciate your feedback, though! :)

**Comment by DeedleFake** [comment]
I read through the issue but there doesn't seem to be any mention of what the insides of the functions would have access to. For example, let's say you have a button in Flutter that has an `onPressed` that needs access to the current `context`. Would this be possible?

```dart
class Example extends StatelessWidget {
  const Example({Key? key}) : super(key: key);

  void _pressed(BuildContext context) {
    // Do something.
  }

  @override
  Widget build(BuildContext context) => const ElevatedButton( // Closure is constant because of const use here.
    child: Text('Do Something'),
    onPressed: () => _pressed(context), // Closes over the context variable.
  );
}
```

If this is possible, then there _would_ actually be something that `const` closures could do that the current workaround of top-level functions can't. I'm not familiar enough with Dart's implementation to know if this _is_ actually possible, but I do know that Go has managed to do it. Closures in Go just compile straight to normal top-level functions with a special hidden extra argument that is used to pass anything that was implicitly accessed. In other words, a function literal appearing at a specific point in the code becomes its own top-level function, and when the reference to that function provided by the literal is called it automatically bundles in the things that were accessed from the parent scope.

**Comment by lrhn** [comment]
A constant function value would definitely not have access to any dynamic state. It would only be able to access static declarations and other constant values.

**Comment by DeedleFake** [comment]
I was under the impression that this proposal was to make the actual function reference itself constant, not the body of the function. Calling a function would still be a runtime operation, I assume. Honestly, it might be possible to just make _all_ function literals constant by default.

Functions that can be called as a constant expression would be neat as well, but I think that would be more like `async` than something specific to function literals. For example,

```dart
doSomething(
  onSomething: const (v) {
    // The body of the function is completely normal. In other words, there's no difference
    // between this function and a top-level or static function because under the hood it
    // actually is one.
    print(v);
  },
);
```

vs.

```dart
// The body of the function is restricted to constant expressions.
int add(int a, int b) const => a + b;

const thisWouldThusBeValid = add(1, 2);
```

**Comment by lrhn** [comment]
The body of the function need not be constant expressions, but it cannot refer to something that depends on runtime *scopes*.
Constant expressions evaluate to *one* value, every time they are evaluated. That function's body cannot refer to any variable in scope which isn't the same variable on *every* evaluation.

It can refer to a static variable, because there is only one instance of that variable.
It cannot refer to a local variable declared outside of the const function itself (it can refer to its own parameters and local variables just fine), and it cannot refer to `this` (if created in a scope where there might be a `this`), because there isn't *one* such variable, there is one variable per invocation of the surrounding code. There is only one constant function value.


I'm not particularly worried about canonicalizing function literals that *can* be constant. Being able to distinguish functions which do *exactly* the same isn't a big issue.

**Comment by tatumizer** [comment]
> It can refer to a static variable, because there is only one instance of that variable.

Then maybe "static", rather than "const", would be a more fitting word for these functions?

**Comment by lrhn** [comment]
Yes, the functions themselves will effectively be static functions. The *function literal expression* will be a constant expression.
It's "constant (function literals)" that's being asked for, not "(constant function) literals".

(And the implementation plan, and existing workaround, is to hoist the function literal into a freshly-named static function declaration in the same static scope, then refer to it by name. That's a constant expression evaluating to a single canonicalized static function.)

**Comment by tatumizer** [comment]
The problem with calling it "const" would be that the users might be tempted to write it not only for an anonymous closure but for the named functions, too:
```dart
const f(int x, int y) => x+y;
```
and then expect this function to be callable in a "const" context:
```dart
const z=f(1, 2);
```
which won't work. Here, two potential interpretations of the function constness will collide, and dart will be drawn into the discussion of "constant (function literals)" vs "(constant function) literals". This could be avoided simply by using a different word ("static").

**Comment by lrhn** [comment]
*Declaring* a local function as `const` is actually reasonable. It'd be like `const f = const (int x, int y) => x + y;`, except that it can also call itself recursively.

If people start thinking they can call functions as part of constant evaluation, they'll figure out soon enough that it won't fly.
(Unless we introduce real "constant functions" too, which is possible, but ... probably not likely any time soon.)

But that's a marketing issue.  The first question is whether we can, and want to, introduce constant function-literal expressions./

**Comment by tatumizer** [comment]
Let's look at the issue from a different perspective. Suppose a function is able to declare local-static vars:
```dart
void foo() {
  static invocationCounter=0;
  invocationCounter++;
  print("foo was called $invocationCounter times");
}
```
We certainly have to use the keyword "static" here rather than const (it's not const at all).
My point is that your example of local function `f` is much closer to the same category as `invocationCounter` than it is to the "constness" category. As soon as we decide to support "local-static" declarations, this will automatically mean that we support both local-static functions and local-static variables, so it's a more general (and less ad-hoc) concept than the const alternative. Agree?

**Comment by lrhn** [comment]
I'd be happy with static local declarations too. The `const` function literals are needed for the places where a *constant* *literal* is needed, which is, honestly, pretty much just default values.

**Comment by lucavenir** [comment]
I was confused about `Function`s not being `const`. Indeed, functions _must_ be defined at _compile time_, so having them _not_ `const` sounds confusing when learning Dart.

This said, how will `Dart 3` relate to this issue? Is it linked with static metaprogramming and/or other features? Is it going to be addressed by then or it's still a low priority?

Thanks for your hard work, Dart team 💙

**Comment by eernstg** [comment]
@lucavenir, thanks for the kind words to the team!

> functions _must_ be defined at _compile time_

That's not quite true. For instance:

```dart
typedef F = int Function();

F g(int x) => () => x;

void main() {
  F f1 = g(2);
  F f2 = g(3);
  print(f1()); // '2'.
  print(f2()); // '3'.
}
```

In this example, we couldn't make `() => x` a constant expression. We don't even know how it should behave before we have chosen to consider a specific invocation of `g`, and there could be many invocations of `g` in the same program execution. So we cannot possibly create a single object representing the value of `() => x` at compile time, and then return that same object every time `g` is invoked.

That's the reason why proposals about constant function literals always have to put some constraints on the body of that function (e.g., that the body must not contain any references to any non-global declarations).

**Comment by jodinathan** [comment]
>> That's the reason why proposals about constant function literals always have to put some constraints on the body of that function (e.g., that the body must not contain any references to any non-global declarations).

isn't that how `String` and `int` works already?  
by reading the compiled JS it seems that the compiler knows when a raw type variable is const even when not specified, ie:

```dart
final foo = 'foo';
final bar = '$foo bar';

print(bar);
```

compiles to something like

```js
console.log('foo bar');
```

same with numbers:

```dart
final foo = 5;
final bar = 3 + foo;

print(bar); // compiles to console.log(8);
```

**Comment by eernstg** [comment]
It would be possible to run the const-related checks on every function literal, and then consider each of them a constant expression if it satisfies the constraints. However, it might be easier to read the code, and to maintain it, if a function literal is only a constant expression when (1) there is a `const` modifier in front of it, or it occurs in a constant context, and (2) it satisfies those const-related constraints.

**Comment by lucavenir** [comment]
> That's not quite true. For instance:
That's embarassing, I totally forgot about higher order functions 😅

> That's the reason why proposals about constant function literals always have to put some constraints on the body of that function (e.g., that the body must not contain any references to any non-global declarations).
So it's doable, as long as no parameters are giving, nor references to non-const global declarations are involved, right?

**Comment by eernstg** [comment]
It's certainly doable. I think the main reason why it hasn't been given a very high priority is that when a function literal is const-able, it can also be lifted:

```dart
void main() {
  const f1 = (String s) => 2; // Could be allowed, if constant function literals were supported.
  const fs1 = [f1]; // Could then use `f1` in other constant expressions.

  const fs2 = [f2]; // But we can do this today---same thing.
}

int f2() => 2;
```

**Comment by GregoryConrad** [comment]
Would this feature also support compile-time evaluation/execution of these constant function literals at compile time (like Rust's `const fn`), assuming they are pure? Or is that a separate issue entirely?

It'd be nice to be able to use a const pure function to instantiate `const` objects when this is not possible through just a constructor.

**Comment by mraleph** [comment]
@GregoryConrad what you are asking for is a larger feature which expands the set of expression / statements which compiler can evaluate in constant context. See all other issues tagged with [`enhanced-const`](https://github.com/dart-lang/language/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aenhanced-const)

**Comment by shtse8** [comment]
could it be possible to please const func in annotation?

**Comment by LewisHolliday** [comment]
I often find myself wanting to write this pattern:

```dart
class Widget {
  const Widget(void Function()? f)
      : onPressed = f == null ? null : (() {
          // do something here like logging
          f();
        }); // Invalid constant value.dart(invalid_constant)

  final void Function()? onPressed;
}
```
Unfortunately, the above is invalid. I'm not sure if there's a practical workaround that doesn't burden the user who instantiates `Widget`.

**Comment by TekExplorer** [comment]
The workaround for that is to just have a method or getter in your widget that does the wrapping. let the onPressed be passed in directly.
```dart
class SomeWidget ... {
  const SomeWidget({this.onPressed})
  final VoidCallback? onPressed;

  VoidCallback? get _onPressed {
    if (onPressed == null) return null;
    // do whatever
    return () {
      onPressed();
    }
  }
  Widget build() {...}
} 
```
wrapping a function like that in the constructor doesnt really provide much benefit at all.

**Comment by LewisHolliday** [comment]
The point is to allow using the constructor argument `onPressed` as a constant. I should have been more clear in my original example. I've edited my original comment. 
The following is another pattern that is solely blocked by this issue:
```dart
import 'package:flutter/material.dart';

class LoggingInkWell extends InkWell {
  const LoggingInkWell({
    super.key,
    required Function()? onTap,
    super.child,
    super.borderRadius,
  }) : super(onTap: onTap == null ? null : () {
    // Do something such as logging here.
    onTap();
  }); // Error: Invalid constant value
}
```
The function literal in the false branch of the ternary could be const if Dart allowed constant literals.

**Comment by TekExplorer** [comment]
That still wouldnt work, because `onTap` is not const.
Now, yes. it would be if the const constructor was used, but currently that does not transfer.
see #2000

**Comment by renatoathaydes** [comment]
It would be nice to be able to take a const reference to a const's method:

```dart
class MyConst {
  const MyConst();
  int foo() => 1;
}

const myFoo = const MyConst().foo;

main() {
  print(myFoo());
}
```

This fails with `Const variables must be initialized with a constant value.`.

I think this may be the same issue as this one?

**Comment by rrousselGit** [comment]
How would this work when applying hot-reload?

Say a user defines:

```dart
const a = () => 'hello';
```

And changes the function to `const a = () => 'bar')`.

Would:
- old `a` be _identical_ to new `a`, and the function be updated to the new behavior (like functions)
- or would `a` contain a new function, so old `a` != new `a`, like when changing `const a = 42` to `const a = 21`

If the latter, what about:

```dart
const a = () => 'hello';
const b = () => a();
```

If we change the source of `a` and a new function instance is made, would `b` get a new instance too?

**Comment by Number-3434** [comment]
I should say that const methods can only refer to final fields on the declaring class, if the class was created as const.

Also, they can refer to other const methods, and other constants, but cannot use any assignment operators

**Comment by lrhn** [comment]
@Number-3434 
That sounds more like C++ const functions.
Dart constant functions, which already exist since all top-level and static functions are constants, can do anything at runtime, but cannot close over any object instances, and cannot have a `this` reference.
They can refer to any static declaration directly, and to any other object through their parameters or through global variables, but they cannot refer directly to instance members of the surrounding class, since there is no insurance to access them through.
And there is no restriction on what the body code can do, including doing assignments.

It's just not that kind of `const`.

**Comment by Number-3434** [comment]
> Dart constant functions, which already exist since all top-level and static functions are constants

Does that mean that you could (in theory) stick the `const` keyword in front of the method?

Would this be rather simple to achieve?

**Comment by philer** [comment]
12 years…

At this point the total time that people have wasted learning out how to work around this must exceed the time it would have taken to implement by several magnitudes.

**Comment by TekExplorer** [comment]
Constant closures would be great.

Seriously, with metaprogramming coming around the corner, and existing annotations/builders, it would greatly enhance usage.

**Comment by TekExplorer** [comment]
You say that, but not all annotations relevant to macros will themselves be macros.

**Comment by wy2010344** [comment]
Instead of using const or some other modifier, how about just making this function a private function at the top level of the current file? I just don't want to name it.
